### PR TITLE
[R-package] raise an informative error when custom objective produces incorrect output (fixes #5323)

### LINKED ISSUE #5323: [R-package] Provide more informative error when custom objective function returns data with incorrect shape
## Description
I try to run a custom loss function but get the error: 
booster$update(fobj = fobj) : No object function provided

## Reproducible example
not completely reproducible in terms of entire data, but the code looks like this

```
MAELoss<- function(preds, dtrain) {
  labels <- attr(dtrain, 'label')
  grad <- abs(preds - labels)
  hess <- rep(0,length(preds))
  return(list(hess = hess,grad = grad))
}


 model <- lgb.train(modelparams
                    ,LGBData
                    ,nthread = 8
                    , verbose= 1
                    , eval_freq = VerboseFreq
                    ,valids=list(OOF=LGBOOF,Train=LGBData) 
                    ,categorical_feature = CategoricalFeats
                    ,early_stopping_rounds = EarlyStoppingN
                    ,nrounds = NumRounds
                    ,obj = MAELoss
                   )
```

## Environment info
lgbm 3.3.2
install.packages("lightgbm")

Windows 10
R 4.2

if i change the 'MAELoss' to "regression"
it works ok, it just won't accept my function.

Thanks!

### PR DESCRIPTION
**Pull request by jmoralez**
Adds a check to verify that the shapes of grad and hess are correct (same as the number of predictions) to raise an informative error when they're not. This is similar to the one added in #4815 for the python-package.

Fixes #5323.

### CODE REVIEW

File: R-package/R/lgb.Booster.R [hunk 1]
```diff
@@ -231,21 +231,33 @@ Booster <- R6::R6Class(
...
           stop("lgb.Booster.update: custom objective should
             return a list with attributes (hess, grad)")
         }
 
+        # Check grad and hess have the right shape
+        n_grad <- length(gpair$grad)
+        n_hess <- length(gpair$hess)
+        n_preds <- length(preds)
+        if (n_grad != n_preds) {
+          stop(sprintf("grad should have length %d, got %d", n_preds, n_grad))
```
Comment (jameslamb) [inline R-package/R/lgb.Booster.R:248]:
```suggestion
          stop(sprintf("Expected custom objective function to return grad with length %d, got %d.", n_preds, n_grad))
```

Can we please modify this error message (and the one about the hessian) to make it clearer to users that their custom objective function's implementation is the most likely cause?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Very nice, thanks for this!

I'm not worried about the performance overhead of doing this on every iteration, since:

* `length()` should be O(1)
* users already should expect training with custom R code to be slower than using the built-in objectives

I just left one suggestion about the language in the error messages.

**Review by jameslamb — APPROVED** [review]
works for me! thanks
